crates/sui-framework/packages/sui-framework/sources/accumulator.move

```diff
@@ -69,6 +71,142 @@ fun root_remove_accumulator<K, V: store>(accumulator_root: &mut AccumulatorRoot,
+fun accumulator_root_owner_exists(accumulator_root: &AccumulatorRoot, owner: address): bool {
+    dynamic_field::exists_with_type<OwnerKey, Owner>(&accumulator_root.id, OwnerKey { owner })
+}
+
+use fun accumulator_root_owner_exists as AccumulatorRoot.owner_exists;
+
+/// Borrow an owner object mutably.
+fun accumulator_root_borrow_owner_mut(
+    accumulator_root: &mut AccumulatorRoot,
+    owner: address,
+): &mut Owner {
+    dynamic_field::borrow_mut(&mut accumulator_root.id, OwnerKey { owner })
+}
+
+use fun accumulator_root_borrow_owner_mut as AccumulatorRoot.borrow_owner_mut;
+
+/// Attach an owner object to the accumulator root.
+fun accumulator_root_attach_owner(accumulator_root: &mut AccumulatorRoot, owner: Owner) {
+    dynamic_field::add(&mut accumulator_root.id, OwnerKey { owner: owner.owner }, owner);
+}
+
+use fun accumulator_root_attach_owner as AccumulatorRoot.attach_owner;
+
+/// Detach an owner object from the accumulator root.
+fun accumulator_root_detach_owner(accumulator_root: &mut AccumulatorRoot, owner: address): Owner {
+    dynamic_field::remove(&mut accumulator_root.id, OwnerKey { owner })
+}
+
+use fun accumulator_root_detach_owner as AccumulatorRoot.detach_owner;
+
+/// === Metadata functions ===
+
+/// Create a metadata object for a new balance object with type T.
+/// The metadata will be attached to the owner object `owner`.
+/// If the owner object does not exist, it will be created.
+fun create_accumulator_metadata<T>(
+    accumulator_root: &mut AccumulatorRoot,
+    owner: address,
+    ctx: &mut TxContext,
+) {
+    let metadata = Metadata<T> {
+        fields: bag::new(ctx),
+    };
+
+    if (accumulator_root.owner_exists(owner)) {
+        let accumulator_owner = accumulator_root.borrow_owner_mut(owner);
+        assert!(accumulator_owner.owner == owner, EInvariantViolation);
+        accumulator_owner.attach_metadata(metadata);
+    } else {
+        let mut accumulator_owner = Owner {
+            balances: bag::new(ctx),
+            owner,
+        };
+        accumulator_owner.attach_metadata(metadata);
+        accumulator_root.attach_owner(accumulator_owner);
+    }
+}
+
+use fun create_accumulator_metadata as AccumulatorRoot.create_metadata;
+
+/// Remove the metadata object for a balance object with type T.
+/// The metadata will be detached from the owner object `owner`.
+/// If there are no more balance objects attached to the owner object,
+/// the owner object will be destroyed.
+fun accumulator_metadata_remove<T>(accumulator_root: &mut AccumulatorRoot, owner: address) {
+    let is_empty = {
+        let accumulator_owner = accumulator_root.borrow_owner_mut(owner);
+        let Metadata { fields } = accumulator_owner.detach_metadata<T>();
+        fields.destroy_empty();
+        accumulator_owner.balances.is_empty()
+    };
+
+    if (is_empty) {
+        let owner = accumulator_root.detach_owner(owner);
+        owner.destroy();
+    }
```

Comment:
Mostly stared at this because of the shadowing of `owner` so just for consideration
```suggestion
    if (is_empty) accumulator_root.detach_owner(owner).destroy();
```